File: DesignPatterns/Behavioral/Iterator/Java/ChannelClient.java
import java.util.ArrayList;
import java.util.List;

enum ChannelType {
  ENGLISH, HINDI, TAMIL, ALL;
}

class Channel {
  private double freq;
  private ChannelType type;

  public Channel(double freq, ChannelType type) {
    this.freq = freq;
    this.type = type;
  }

  public double getFreq() {
    return this.freq;
  }

  public ChannelType getType() {
    return this.type;
  }

  public String toString() {
    return "Frequency = " + this.freq + ", Type = " + this.type;
  }
}

interface ChannelCollection {
  public void addChannel(Channel c);
  public void removeChannel(Channel c);
  public ChannelIterator createIterator(ChannelType type);
}

interface ChannelIterator {
  public boolean hasNext();
  public Channel next();
}

class ChannelCollectionImpl implements ChannelCollection {
  private List<Channel> channelList;

  public ChannelCollectionImpl(){
    channelList = new ArrayList<Channel>();
  }

  public void addChannel(Channel c) {
    this.channelList.add(c);
  }

  public void removeChannel(Channel c) {
    this.channelList.remove(c);
  }

  public ChannelIterator createIterator(ChannelType type) {
    return new ChannelIteratorImpl(type, this.channelList);
  }
  //The inner class implementation of iterator prevents the channel iterator to be used by any other collections
  private class ChannelIteratorImpl implements ChannelIterator {
    private ChannelType type;
    private List<Channel> channels;
    private int position;

    public ChannelIteratorImpl(ChannelType type, List<Channel> channelList) {
      this.type = type;
      this.channels = channelList;
    }

    public boolean hasNext() {
      while(position < channels.size()){
        Channel c = channels.get(position);
        if (c.getType().equals(type) || type.equals(ChannelType.ALL)) {
          return true;
        } else {
          position++;
        }
      }
      return false;
    }

    public Channel next() {
      Channel c = channels.get(position);
      position++;
      return c;
    }

  }
}

public class ChannelClient {
  public static void main(String[] args) {
    ChannelCollection cc = populateChannels();
    ChannelIterator ciAll = cc.createIterator(ChannelType.ALL);
    while(ciAll.hasNext()) {
      System.out.println(ciAll.next());
    }
    System.out.println("****************");
    ChannelIterator ciEng = cc.createIterator(ChannelType.ENGLISH);
    while(ciEng.hasNext()){
      System.out.println(ciEng.next());
    }
  }

  private static ChannelCollection populateChannels() {
    ChannelCollection cc = new ChannelCollectionImpl();
    cc.addChannel(new Channel(98.5, ChannelType.ENGLISH));
    cc.addChannel(new Channel(99.5, ChannelType.HINDI));
    cc.addChannel(new Channel(100.5, ChannelType.TAMIL));
    cc.addChannel(new Channel(101.5, ChannelType.ENGLISH));
    cc.addChannel(new Channel(102.5, ChannelType.HINDI));
    cc.addChannel(new Channel(103.5, ChannelType.TAMIL));
    cc.addChannel(new Channel(104.5, ChannelType.ENGLISH));
    cc.addChannel(new Channel(105.5, ChannelType.HINDI));
    cc.addChannel(new Channel(106.5, ChannelType.TAMIL));
    return cc;
  }
}

File: DesignPatterns/Behavioral/Iterator/Ruby/channel_client.rb
module Constants
  ENGLISH = "ENGLISH".freeze
  TAMIL = "TAMIL".freeze
  HINDI = "HINDI".freeze
  ALL = "ALL".freeze
end

class Channel
  attr_reader :freq, :type

  def initialize(freq, type)
    @freq = freq
    @type = type
  end

  def to_s
    "Frequency = #{@freq}, Type = #{@type}"
  end
end

class Iterator
  attr_reader :pos

  def initialize
    @pos = 0
  end
end

class ChannelIterator < Iterator
  attr_reader :channels, :type

  def initialize(channels, type)
    super()
    @channels = channels.clone
    @type = type
  end

  def next
    value = @channels[@pos]
    @pos += 1
    value
  end

  def has_next?
    while @pos < @channels.length
      value = @channels[@pos]
      if @type == Constants::ALL || value.type == @type
        return true
      else
        @pos += 1
      end
    end
    return false
  end
end

class Collection
  attr_reader :items, :total_items

  def initialize
    @items = []
    @total_items = 0
  end
end

class ChannelCollection < Collection
  def add_item(item)
    @items << item
  end

  def create_iterator(type)
    ChannelIterator.new(@items, type)
  end
end

cc = ChannelCollection.new
cc.add_item(Channel.new(98.5, Constants::ENGLISH))
cc.add_item(Channel.new(99.5, Constants::TAMIL))
cc.add_item(Channel.new(101.5, Constants::ENGLISH))
cc.add_item(Channel.new(103.5, Constants::HINDI))
cc.add_item(Channel.new(96.5, Constants::TAMIL))
cc.add_item(Channel.new(106.5, Constants::HINDI))

eng_iterator = cc.create_iterator(Constants::ENGLISH)
while eng_iterator.has_next? do
  puts eng_iterator.next
end

puts "*********"

all_iterator = cc.create_iterator(Constants::ALL)
while all_iterator.has_next? do
  puts all_iterator.next
end
